clippy_lints/src/transmute/mod.rs

```diff
@@ -0,0 +1,363 @@
+    /// // different size, therefore likely out-of-bounds memory access
+    /// // You absolutely do not want this in your code!
+    /// unsafe {
+    ///     std::mem::transmute::<_, Vec<u32>>(vec![2_u16])
+    /// };
+    /// ```
+    ///
+    /// You must always iterate, map and collect the values:
+    ///
+    /// ```rust
+    /// vec![2_u16].into_iter().map(u32::from).collect::<Vec<_>>();
+    /// ```
+    pub UNSOUND_COLLECTION_TRANSMUTE,
+    correctness,
+    "transmute between collections of layout-incompatible types"
+}
+
+declare_lint_pass!(Transmute => [
+    CROSSPOINTER_TRANSMUTE,
+    TRANSMUTE_PTR_TO_REF,
+    TRANSMUTE_PTR_TO_PTR,
+    USELESS_TRANSMUTE,
+    WRONG_TRANSMUTE,
+    TRANSMUTE_INT_TO_CHAR,
+    TRANSMUTE_BYTES_TO_STR,
+    TRANSMUTE_INT_TO_BOOL,
+    TRANSMUTE_INT_TO_FLOAT,
+    TRANSMUTE_FLOAT_TO_INT,
+    UNSOUND_COLLECTION_TRANSMUTE,
+    TRANSMUTES_EXPRESSIBLE_AS_PTR_CASTS,
+]);
+
+impl<'tcx> LateLintPass<'tcx> for Transmute {
+    #[allow(clippy::similar_names, clippy::too_many_lines)]
+    fn check_expr(&mut self, cx: &LateContext<'tcx>, e: &'tcx Expr<'_>) {
+        if_chain! {
+            if let ExprKind::Call(ref path_expr, ref args) = e.kind;
+            if let ExprKind::Path(ref qpath) = path_expr.kind;
+            if let Some(def_id) = cx.qpath_res(qpath, path_expr.hir_id).opt_def_id();
+            if match_def_path(cx, def_id, &paths::TRANSMUTE);
+            then {
+                // Avoid suggesting from/to bits and dereferencing raw pointers in const contexts.
+                // See https://github.com/rust-lang/rust/issues/73736 for progress on making them `const fn`.
+                // And see https://github.com/rust-lang/rust/issues/51911 for dereferencing raw pointers.
+                let const_context = in_constant(cx, e.hir_id);
+
+                let from_ty = cx.typeck_results().expr_ty(&args[0]);
+                let to_ty = cx.typeck_results().expr_ty(e);
+
+                // If useless_transmute is triggered, the other lints can be skipped.
+                if useless_transmute::check(cx, e, from_ty, to_ty, args) {
+                    return;
+                }
+
+                let triggered = wrong_transmute::check(cx, e, from_ty, to_ty) ||
+                crosspointer_transmute::check(cx, e, from_ty, to_ty) ||
+                transmute_ptr_to_ref::check(cx, e, from_ty, to_ty, args, qpath) ||
+                transmute_int_to_char::check(cx, e, from_ty, to_ty, args) ||
+                transmute_ref_to_ref::check(cx, e, from_ty, to_ty, args, const_context) ||
+                transmute_ptr_to_ptr::check(cx, e, from_ty, to_ty, args) ||
+                transmute_int_to_bool::check(cx, e, from_ty, to_ty, args) ||
+                transmute_int_to_float::check(cx, e, from_ty, to_ty, args, const_context) ||
+                transmute_float_to_int::check(cx, e, from_ty, to_ty, args, const_context) ||
+                unsound_collection_transmute::check(cx, e, from_ty, to_ty);
```

Comment:
Just a question.
These lines seem to implicitly define the priority order of the lints.
For example, `useless_transmute` > `others` and  `wrong_transmute` > `crosspointer_transmute` and so on.

Is it ok? I think this may become a problem if a new lint has a mutual condition that triggers both the lint and another lint.

Reply:
Good point. 
Generally, all the lints here are exclusive; if a code violates `transmute_int_to_char`, then this code does _never_ trigger `transmute_ptr_to_ptr`, `transmute_int_to_bool`, and so on.
But there are a few exceptions. For example, `useless_transmute` and `transmute_ref_to_ref` can potentially be triggered at the same time - this is NOT exclusive. In such a case, I think we'd prefer one diagnostic over multiple ones to avoid getting the programmer overwhelmed. This is why the lint priority is necessary. (Also... the test cases would fail if we removed the priority :-|)

If a new lint related to transmute is added in the future, we will have to consider the priority of it to put it in the appropriate position. If anyone comes up with a better way, that would be really welcome.

Reply:
> In such a case, I think we'd prefer one diagnostic over multiple ones to avoid getting the programmer overwhelmed. 

I agree with you in almost all cases. My concern here is that if `#[allow(clippy::useless_transmute)]` is specified by the user, then `transmute_ref_to_ref` may also be suppressed unexpectedly.

I feel like this is a problem that should be generalized through `clippy`.

Reply:
> My concern here is that if `#[allow(clippy::useless_transmute)]` is specified by the user, then `transmute_ref_to_ref` may also be suppressed unexpectedly.

Ah that makes much sense. I guess the current master branch also has that problem; once the lint checking flow goes into `useless_transmute`, the other lints won't never be checked even if `#[allow(clippy::useless_transmute)]` is specified.
With this fact in mind, I feel like prioritizing the lints might be bad.

Reply:
I'd say it would be better to keep the current behavior in this PR by confirming the existing test cases are passing, because the main focus of this PR is refactoring. We can consider further how to address the priority issue in another issue or PR.

Reply:
> the main focus of this PR is refactoring. 

I agree with you! I created the topic to discuss this issue on [zulip](https://rust-lang.zulipchat.com/#narrow/stream/257328-clippy/topic/Priority.20order.20of.20lints/near/225983716)

Reply:
If two lints **cannot** conflict, I would avoid short circuiting them like this. Only if one lint triggering should depend on another lint this should be done. So I would write this as

```rust
let mut linted = check_lint_A();
linted |= check_lint_B(); // no dependency
linted |= check_lint_C() || check_lint_D(); // D depends on C
...

if !linted {
    check_lint_XYZ(); // XYZ depends on ALL lints
}
```

Reply:
@flip1995 Looks cool. I'll rewrite the code!

Reply:
@flip1995 Rewriting and rebasing with master are done. PTAL :)